File: examples/java-gradle/src/main/java/com/example/abstracts/AbstractEntity.java
package com.example.abstracts;

import com.example.interfaces.JsonSerializable;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;

/**
 * Abstract base class for all entities.
 */
public abstract class AbstractEntity implements JsonSerializable {

    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final ObjectMapper PRETTY_MAPPER = new ObjectMapper()
            .enable(SerializationFeature.INDENT_OUTPUT);

    protected Long id;

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public boolean isNew() {
        return id == null || id == 0;
    }

    @Override
    public String toJson() {
        try {
            return MAPPER.writeValueAsString(this);
        } catch (JsonProcessingException e) {
            return "{}";
        }
    }

    @Override
    public String toPrettyJson() {
        try {
            return PRETTY_MAPPER.writeValueAsString(this);
        } catch (JsonProcessingException e) {
            return "{}";
        }
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        AbstractEntity that = (AbstractEntity) obj;
        return id != null && id.equals(that.id);
    }

    @Override
    public int hashCode() {
        return id != null ? id.hashCode() : 0;
    }
}

File: examples/kotlin-gradle/src/main/kotlin/com/example/abstracts/AbstractEntity.kt
package com.example.abstracts

import com.example.interfaces.JsonSerializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/**
 * Abstract base class for all entities.
 */
abstract class AbstractEntity : JsonSerializable {

    abstract var id: Long?

    val isNew: Boolean
        get() = id == null || id == 0L

    companion object {
        val json = Json { prettyPrint = false }
        val prettyJson = Json { prettyPrint = true }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        other as AbstractEntity
        return id != null && id == other.id
    }

    override fun hashCode(): Int = id?.hashCode() ?: 0
}
